slapo/op/flash_attention_triton.py

```diff
@@ -0,0 +1,1032 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Copy from:
+https://github.com/jfc4050/fast-ops/blob/main/fast_ops/flash_attention/flash_attention_triton.py
+
+*Experimental* implementation of FlashAttention in Triton.
+There's still some unresolved race conditions for certain input types.
+In particular there's issues if:
+- seq_len % 128 != 0
+- head_dim != 128 
+before using this please double check the unit tests to make sure your use case
+is covered.
+
+We use the FlashAttention implementation from Tri Dao
+(https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attn_triton.py)
+as a starting point, who in turn used the FlashAttention implementation from Phil Tillet
+(https://github.com/openai/triton/blob/master/python/tutorials/06-fused-attention.py)
+as a starting point.
+
+This version adds dropout support and does some additional optimization, at the
+cost of breaking some input cases that worked in Tri Dao's implementation.
+
+TODOs:
+* upgrade triton version
+* block-sparse attention
+* NestedTensor support
+* bias gradient support
+"""
+
+import math
+from typing import Tuple
+
+import numpy as np
+import torch
+from torch import Tensor
+from torch.autograd import Function
+
+import triton
+import triton.language as tl
```

Comment:
Porting this file may require users to manually install Triton. Do we have error messages or instructions somewhere to inform users to install Triton?